lib/cstruct.ml

```diff
@@ -417,7 +415,11 @@ let of_string ?allocator ?off ?len buf =
 let of_bytes ?allocator ?off ?len buf =
   of_data_abstract blit_from_bytes Bytes.length ?allocator ?off ?len buf
 
-let of_hex str =
+let of_hex ?(off = 0) ?len str =
+  let str =
+    let l = match len with None -> String.length str - off | Some l -> l in
+    String.sub str off l
+  in
   let string_fold ~f ~z str =
     let st = ref z in
     ( String.iter (fun c -> st := f !st c) str  ; !st )
```

Comment:
out of scope, you may want to open a PR

Reply:
why is it out of scope? It is this PR that introduces the copying in String.sub

Reply:
the current change is fairly easy to review and understand, messing around with the `string_fold` may improve the performance (does actually anyone care about the `of_hex` performance? to me this is mainly a function for test cases where you have literal hex-encoded input).

now, your suggested solution uses mutable state (ok, rather local one), which I find hard to reason about, and try to avoid in my code.

as said earlier, i prefer here the easy-to-read over the high-performance solution (and think my time is better spent on other projects & PRs). so no, sorry, I won't alter the proposed PR.